rkt/containers.go

```diff
@@ -0,0 +1,222 @@
+)
+
+var (
+	skipUUID = errors.New("skip this container")
+)
+
+// withContainerHandles iterates over the included containers calling function f for every container.
+func walkContainers(include includeMask, f func(*container)) error {
+	ls, err := getContainers(include)
+	if err != nil {
+		return fmt.Errorf("failed to get containers: %v", err)
+	}
+	sort.Strings(ls)
+
+	for _, uuid := range ls {
+		c, err := getContainer(uuid)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Skipping %q: %v\n", uuid, err)
+			continue
+		}
+
+		// omit garbage containers if garbage wasn't included
+		// this is to cover a race between getContainers finding the uuids and `rkt gc` renaming a container
+		if c.isGarbage && include&includeGarbageDir == 0 {
+			c.Close()
+			continue
+		}
+
+		f(c)
+		c.Close()
+	}
+	return nil
+}
+
+// getContainer returns a container struct representing the given container.
+// The returned lock is always left in an open but unlocked state.
+// The container must be closed using container.Close()
+func getContainer(uuid string) (*container, error) {
+	c := &container{
+		uuid:       uuid,
+		isGarbage:  true,
+		isExited:   true,
+		isDeleting: false,
+	}
+
+	l, err := lock.NewLock(filepath.Join(garbageDir(), uuid))
+	if err == lock.ErrNotExist {
+		l, err = lock.NewLock(filepath.Join(containersDir(), uuid))
+		c.isGarbage = false
+		c.isExited = false
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error opening container %q: %v", uuid, err)
+	}
+
+	if err = l.TrySharedLock(); err != nil {
+		if err != lock.ErrLocked {
+			l.Close()
+			return nil, fmt.Errorf("unexpected lock error: %v", err)
+		}
+		if c.isGarbage {
+			c.isDeleting = true
+		}
+	} else {
+		l.Unlock()
+		c.isExited = true
+	}
+
+	c.DirLock = l
+	cfd, err := c.Fd()
+	if err != nil {
+		c.Close()
+		return nil, fmt.Errorf("error getting lock fd: %v", err)
+	}
+
+	var st syscall.Stat_t
+	if err := syscall.Fstat(cfd, &st); err != nil {
+		return nil, fmt.Errorf("error stating container: %v", err)
+	}
+
+	// The container directory's mtime is approximately the time it was created
+	c.createdAt = time.Unix(st.Mtim.Unix())
+
+	// If the container is garbage, its Ctim reflects when it was marked as such
+	if c.isGarbage {
+		c.garbageAt = time.Unix(st.Ctim.Unix())
+	}
+
+	return c, nil
+}
+
+// path returns the path to the container where it was opened.
+func (c *container) path() string {
+	if c.isGarbage {
+		return c.garbagePath()
+	}
+	return c.containersPath()
+}
+
+// garbagePath returns the path to the container where it would be in the garbageDir.
+func (c *container) garbagePath() string {
+	return filepath.Join(garbageDir(), c.uuid)
+}
+
+// containersPath returns the path to the container where it would be in the containersDir.
+func (c *container) containersPath() string {
+	return filepath.Join(containersDir(), c.uuid)
+}
+
+// getContainers returns a list of container uuids in string form.
+func getContainers(include includeMask) ([]string, error) {
```

Comment:
listContainers would be better, since getContainer does something very different now